### PR TITLE
Change code using fixed subset name in HPO

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
This PR changes code using fixed subset name in hpo_api.py
In addition to that, I changed HPO code to raise error once one of HPO trials exit abnormally.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/engine/hpo/hpo_api.py [hunk 1]
```diff
@@ -133,8 +133,15 @@ def hpo_config(self) -> dict[str, Any]:
 
     @hpo_config.setter
     def hpo_config(self, hpo_config: HpoConfig | None) -> None:
-        train_dataset_size = len(self._engine.datamodule.subsets["train"])
-        val_dataset_size = len(self._engine.datamodule.subsets["val"])
+        train_dataset_size = len(
+            self._engine.datamodule.subsets[self._engine.datamodule.config.train_subset.subset_name],
+        )
```
Comment (vinnamkim) [inline src/otx/engine/hpo/hpo_api.py]:
```suggestion
        train_dataset_size = len(
            self._engine.datamodule.train_dataloader(),
        )
```

Comment (eunwoosh) [reply src/otx/engine/hpo/hpo_api.py]: done https://github.com/openvinotoolkit/training_extensions/pull/3101/commits/39bb69f480047100acf076d8c22efa2ca0b4f721

File: src/otx/engine/hpo/hpo_api.py [hunk 2]
```diff
@@ -133,8 +133,11 @@ def hpo_config(self) -> dict[str, Any]:
 
     @hpo_config.setter
     def hpo_config(self, hpo_config: HpoConfig | None) -> None:
-        train_dataset_size = len(self._engine.datamodule.subsets["train"])
-        val_dataset_size = len(self._engine.datamodule.subsets["val"])
+        train_dataset_size = len(self._engine.datamodule.train_dataloader())
+        if self._engine.datamodule.config.val_subset.subset_name in self._engine.datamodule.subsets:
```
Comment (vinnamkim) [inline src/otx/engine/hpo/hpo_api.py]: Raise an error if it cannot create `val_dataloader`

Comment (eunwoosh) [reply src/otx/engine/hpo/hpo_api.py]: As far as I know, anomaly task doesn't use validation dataset. It's for that.

Comment (vinnamkim) [reply src/otx/engine/hpo/hpo_api.py]: https://github.com/openvinotoolkit/training_extensions/blob/0b572ccdaef03f5541340b7eed62e37f03fc8a5a/src/otx/data/anomaly/anomaly.py#L52-L59

Comment (vinnamkim) [reply src/otx/engine/hpo/hpo_api.py]: Anyway, you must need an evaluation to do HPO. How do you HPO for anomaly tasks without a validation set?

Comment (eunwoosh) [reply src/otx/engine/hpo/hpo_api.py]: I found that anomaly models don't use both val_dataloader and test_dataloader. It seems that anomaly models use only train dataset for training.
HPO refers 'callback_monitor' as metric. So, `train_loss_epoch` is used for anomaly.

**Review by harimkang — DISMISSED** [review]

File: src/otx/engine/hpo/hpo_api.py (hunk 2 above)
Comment (vinnamkim) [reply src/otx/engine/hpo/hpo_api.py]: image
I just examined it with my debugger and can find that the evaluation step exists and it comes from the test subset

**Comment by harimkang** [comment]
Please modify the target with this to the releases/2.0.0 branch.

File: src/otx/engine/hpo/hpo_api.py (hunk 2 above)
Comment (eunwoosh) [reply src/otx/engine/hpo/hpo_api.py]: I removed them. Could you check?

**Comment by eunwoosh** [comment]
> Please modify the target with this to the releases/2.0.0 branch.

I'll open same PR for releases/2.0.0 after it's merged.

**Review by vinnamkim — APPROVED** [review]

**Review by harimkang — APPROVED** [review]
